### ISSUE TITLE
Allow zites to overwrite, overrule or set an custom auth_address.

### ISSUE THREAD

**Issue by pexo-zn** [HelloZeroNet/ZeroNet#2350]
**Is your feature request related to a problem? Please describe.**
 To the best of my knowledge, for a zite developer it is currently impossible to change the auth_address at all. Because of this, added certificates will in the eyes of ZeroNet always be intended to be for the current auth_address(which, per zite, cannot change), and also accepted regardless of the signature contents (see this issue). This makes not only anonymous posting impossible but also the ability to have multiple accounts on one zite, since `fileWrite` will only write files, if the current auth_address is in the valid signers list. With no way to change auth_address this is only true for one directory path(which ends in that auth_address). Possible applications for multiple accounts could be a moderation system(possibly even autonomous ones, through election algorithms), where you would have differently privileged user types, as example:

- Owner
- Moderator
- Volunteer
- User As things currently are, a moderator, for example, must always post with his moderator account.

**Describe the solution you'd like**
An api command to either:

- overwrite: Replace the generated auth_address with custom one, but always default to generated one, if the zite developer for example tried to add an invalid privatekey (only either of them is readable at every point in time)
- overrule: Add the custom auth_address as an new field(both are readable at any point in time).
- set: Replace the generated auth_address permanently with the custom one.

An alternative way to implement this would be to allow have an commend like `setAuthIndex` where that will change the auth_address to the subkey of the specified index, similar to the way `userPublickey` from `CryptMessage` plugin does it. But this would be way less flexible than the solutions mentioned above.

In addition to that, in `certAdd` the ability to specify what privatekey the new certificate, actually belongs to and only to assume the current auth_address as default. Something like: `@cmd "certAdd", ["zeroid.bit", auth_type, auth_privkey, user_name, cert_sign], (res) =>{/**Stuff*/}` Where auth_privkey is, if not specified, the private key of the current auth_address.

**Describe alternatives you've considered**
Using current api commands, trying different zite structures and digging through ZeroNet source code to find a way to alter/change the auth_address. None of these I had success with.

**Additional context**
This issue is mirrored from the [GitCenter ZeroNet repository](http://127.0.0.1:43110/1GitLiXB6t5r8vuU2zC6a8GYj9ME6HMQ4t/repo/issues/view/?1P4w4Rvh4vS9r6G5xTiSh8qEW87EKZtnJB/8@1DfrA2M9Qra6stqT6tzwNMYNweswY54KAC)(you need to install ZeroNet to visit this link)

**Comment by pexo-zn** [comment]
Previous conversation on GitCenter:
- >nofish@zeroid.bit commented on Mar 9 2019
    
    I would prefer the setAuthIndex -like solution as it would make it possible to recover it from the master seed
- >pexo@zeroid.bit commented on Mar 10 2019
    
    >make it possible to recover it from the master seed
    
    If one can remember the index, at least.
    
    Anyway, both are good options in my book, although something to have in mind is that `setAuthIndex` will lock you into, at most, 2,147,483,647(signed integer max value) unique addresses, compared to basically infinite unique addresses you get with the other option. But 2.1 Billion addresses are more then enough anyways.
    
    Also there is a security consideration to be made here, as the option to recover the addresses from the master seed will cause all generated addresses to be compromised should the master seed ever be stolen. Maybe an api command to create a temporary throwaway address would solve that, in case that becomes necessary for a zite developer?
    
    Do you have a rough estimate when you are able to implement this?

**Comment by decentralizedauthority** [comment]
> possibly even autonomous ones, through election algorithms

Maybe this will be useful in the upcoming 2020 U.S. elections :rofl: 

> Maybe an api command to create a temporary throwaway address

I don't think that Bitcoin addresses should be generated for "throw" them away... What if someone after you "throw" the address donate some Bitcoin for you? 

I agree with you about the master_seed, I have 100+ zites and none of the addresses are generated using the master_seed. 

Maybe ask @filips123 or @imachug how this can be done, but my guess is you should use IPFS instead of "throw away" addresses.

**Comment by imachug** [comment]
Donating to ZeroNet addresses is not recommended anyway

**Comment by decentralizedauthority** [comment]
> Donating to ZeroNet addresses is not recommended anyway

Why is that @imachug ? It is a valid Bitcoin address and only the person created the address knows the private key. If the private key is secured than I don't see why not to donate to sites on ZeroNet.

**Comment by purplesyringa** [comment]
This is blocking for me. @HelloZeroNet Do you think I could implement it?

**Comment by HelloZeroNet** [comment]
Idea on implementation:
 - Create a `siteIdentityAdd` function that inserts a new option to `certSelect` dialog with auth address devilered from site index + 1 (asks for identity title when called)

Could this be a solution for your use case?

**Comment by purplesyringa** [comment]
I'm not sure, I think I need to generate lots of IDs so probably inserting all the IDs to `certSelect` will clutter it. What about a separate `Anonymous` option in the dialog that allows the site to update the ID with an API call? It's transparent for the user and easy to use and implement. (of course the option should only be shown if `certSelect` is given a special `show_anonymous` argument).

**Comment by HelloZeroNet** [comment]
Not sure if we need separate item in the certSelect menu for this, we could add a `siteIdentitySet auth_address_id`, (limited to 0-10240 to avoid security issues) + add a field to the sidebar that allow to reset to 0/change it manually.

**Comment by purplesyringa** [comment]
LGTM. Is there any wisdom behind choosing 10240 as a limit? Or is it just a "nice" number? (a good parody of 1.44 MB)

**Comment by HelloZeroNet** [comment]
Not really, maybe it would be better to choose different derivation path. Not sure if pybitcointools support is.

**Comment by purplesyringa** [comment]
AFAIK there's none. Limiting the index to 10000 should work. I'll start implementing it soon.
